design/StableRegions-20230209.md

```diff
@@ -0,0 +1,172 @@
+regions at 8MB each, well over the maximum stable memory size is used
+(256GB compared to 32GB, the limit today)
+
+
+## Design details
+
+### API
+
+Internal region allocator operations:
+
+
+ - `initialize` -- called by the RTS, not by the Motoko developer.
+ - `rebuild` -- also called by the RTS, during an post-upgrade.
+
+
+User-facing region allocator operations:
+
+ - `region_new` -- create a dynamic region.
+ - `region_grow` -- grow region by a specified number of pages.
+ - `region_release`* -- release region and reuse region's page blocks.
+ - `region_load` -- read some data from the region.
+ - `region_store` -- store some data into the region.
+
+
+ - * The `region_release` operation is not part of the MVP, but supporting it
+   is important because it means we can transition quickly to an integration
+   with the ambient Motoko GC if we can support it.
+
+## Internal footprint
+
+The state of the allocator is stored in a combination of stable memory and heap memory.
+
+The stable memory state is sufficient to reconstitute the heap objects
+(`rebuild` operation, described in a subsection below).  That means
+that even if the heap is lost (like during an upgrade that fails) the
+stable memory state can fully describe the state parts that should be
+rebuilt into the heap when the upgrade succeeds.
+
+### stable memory fields
+
+ - total allocated buckets, `Nat16`, max value is `32768`.
+ - total allocated regions, `Nat16`, max value is `32767` (one region is reserved for "no region").
+ - bucket-region table (fixed size, about 131kb).
+ - region-object table (fixed size, about 393kb).
+
+### bucket-region table
+
+ - purpose: relate a bucket ID ("page block ID") to its region (if any) and its position in that region.
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BucketRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding bucket ID.
+
+### region-object table
+
+ - 32768 entries (statically sized).
+ - 12 bytes per entry.
+ - entry type = `RegionObjectEntry { size_in_pages: Nat64; heap_object: Nat32 }`
+ - the location of each entry gives its corresponding region ID.
+ - the heap object layout is rebuilt on upgrade.
+ - the heap object is described in its own subsection.
+
+### region heap object
+
+ - `RegionObject { id : Nat16; capacity : Nat16; vec_ptr: Nat32 }`
+ - 8 bytes per entry.
+ - `vec_ptr` points to a vector with `capacity` slots.
+ - to support load and store quickly, the `vec_ptr` is also held in the heap.
+ - capacity of `vec_ptr` doubles when it grows.
+ - no region has more than 32k page blocks, so a `Nat16` suffices for `capacity`.
+
+### Overview of `rebuild`
+
+When a canister upgrades, its heap is lost and the region heap objects must thus be rebuilt.
+
+Specifically, the region-object table points into the heap, and its heap objects are
+rebuilt by using both the bucket-region table, and the region-object table.
```

Comment:
>  then deserialized by returning a fresh object with mutable fields recovered from the region-object-table.

Your description closely matches what I originally intended, but after reading your other comments, I have another, more refined version of my original idea:
 - The region heap object is stable (is serialized and deserialized as stable data), but
 - Its vector is nulled out on serialization, and rebuilt on upgrade (deserialization), rather than being serialized with the heap object.

I don't know about very strong arguments for this variation over *also* serializing the vectors and then deserializing them, except that it would mean that the validity of preallocated tables are harder to test with the kinds of tests we have now.  If we use them on *every* upgrade (at least initially), then we know they work right just by doing ordinary upgrade tests.

OTOH, if we store redundant data on upgrade and only use the preallocated tables when it fails, it may be harder to test that those cases "work" as expected.

WDYT?

Reply:
I actually think its worth finding out when stable memory will be cheap enough to just store the data there. That might make your life much simpler. Otherwise, to actually support perf and recovery we'll need some mechanism that is cheap but keeps the stable metadata in sync with the heap data.

Reply:
I'd ask @ulan when cheap stable memory is due...